runtime/compiler/optimizer/JProfilingValue.cpp

```diff
@@ -454,168 +481,238 @@ TR_JProfilingValue::addProfilingTrees(
-      // In this case we need to store object in temp slot and replace subbsequent reference to that object
-      // Usually splitting of block takes care of uncommoning but in cases when we have a first reference of these objects in the
-      // TreeTop before we are adding the profiling trees (checkCast, instanceOf) as we dereference the object before checkCast or
-      // instanceOf nodes, we might skip generating NULL test in later nodes. To avoid that, we should store the object as well and
-      // replace subsequent references to it.
-      TR::TreeTop *storeValue = TR::TreeTop::create(comp, insertionPoint, storeNode(comp, value->getFirstChild(), storedObjectSymRef));
-      replaceNode(comp, extendedBlock->getEntry(), storeValue->getNextTreeTop(),
-         value->getFirstChild(), TR::Node::createLoad(value->getFirstChild(), storedObjectSymRef));
-      // We need to store the value we are profiling in temp slot to be used by helper. 
-      quickTest = originalBlock->split(storeValue->getNextTreeTop(), cfg, true, true);
-      TR::TreeTop *storeProfilingValue = TR::TreeTop::create(comp, quickTest->getEntry(), storeNode(comp, value, storedValueSymRef));
-      quickTestLastProfilingTT = storeProfilingValue;
+      TR_PersistentProfileInfo *profileInfo = comp->getRecompilationInfo()->findOrCreateProfileInfo();
+      TR_BlockFrequencyInfo *bfi = TR_BlockFrequencyInfo::get(profileInfo);
+      TR::Node *loadIsQueuedForRecompilation = TR::Node::createWithSymRef(value, TR::iload, 0, comp->getSymRefTab()->createKnownStaticDataSymbolRef(bfi->getIsQueuedForRecompilation(), TR::Int32));
+      TR::Node *checkIfQueueForRecompilation = TR::Node::createif(TR::ificmpeq, loadIsQueuedForRecompilation, TR::Node::iconst(value, -1), mainlineReturn->getEntry());
+      TR::TreeTop *checkIfNeedToProfileValue = TR::TreeTop::create(comp, checkIfQueueForRecompilation);
+      iter->append(checkIfNeedToProfileValue);
+      if (origBlockGlRegDeps != NULL)
+         {
+         TR::Node *exitGlRegDeps = copyGlRegDeps(comp, origBlockGlRegDeps);
+         checkIfQueueForRecompilation->addChildren(&exitGlRegDeps, 1); 
+         }
+      lastBranchToMainlineReturnTT = checkIfNeedToProfileValue;
+      if (trace)
+         traceMsg(comp, "\t\t\tCheck if queued for recompilation test performed in block_%d\n", iter->getNumber());
       }
-   else
+
+   // In case profiling vft , need to do null test for the object.
+   if (addNullCheck)
       {
-      TR::TreeTop *storeValue = TR::TreeTop::create(comp, insertionPoint, storeNode(comp, value, storedValueSymRef));
-      replaceNode(comp, extendedBlock->getEntry(), storeValue->getNextTreeTop(),
-         value, TR::Node::createLoad(value, storedValueSymRef));
-      // Split after the store
-      quickTest = originalBlock->split(storeValue->getNextTreeTop(), cfg, true, true);
-      quickTestLastProfilingTT = quickTest->getEntry();
+      /**
```

Comment:
I don't think this should be doxygen

Reply:
Fixed in https://github.com/eclipse/openj9/pull/6585/commits/4babc9149b99cc98d9703d5340c8c21f81c918da